design/configuration-crd.md

```diff
@@ -0,0 +1,271 @@
+
+This design proposes two new CRDs, one that represents a `ContourConfiguration` (Short name `ContourConfig`) and another which represents a `ContourDeployment`.
+The Contour configuration mirrors how the configmap functions today. 
+A Contour Deployment is managed by a controller (aka Operator) which uses the details of the CRD spec to deploy a fully managed instance of Contour inside a cluster.
+
+## Benefits
+- Eliminates the need to translate from a CRD to a configmap (Like the Operator does today)
+- Allows for a place to surface information about configuration errors - the CRD status, in addition to the Contour log files
+- Allows the Operator workflow to match a non-operator workflow (i.e. you start with a Contour configuration CRD)
+- Provides better validation to avoid errors in the configuration file
+- Dynamic restarting of Contour when configuration changes
+
+## Contour Configuration
+
+The Contour configuration file wil be migrated into a Contour Configuration CRD named `Contour`.
+The current config file looks like this:
+
+```yaml
+server:
+  xds-server-type: contour
+gateway:
+  controllerName: projectcontour.io/projectcontour/contour
+incluster: true
+kubeconfig: /path/to/.kube/config
+disableAllowChunkedLength: false
+disablePermitInsecure: false
+tls:
+  minimum-protocol-version: "1.2"
+  cipher-suites:
+  - '[ECDHE-ECDSA-AES128-GCM-SHA256|ECDHE-ECDSA-CHACHA20-POLY1305]'
+  - '[ECDHE-RSA-AES128-GCM-SHA256|ECDHE-RSA-CHACHA20-POLY1305]'
+  - 'ECDHE-ECDSA-AES256-GCM-SHA384'
+  - 'ECDHE-RSA-AES256-GCM-SHA384'
+fallback-certificate:
+  name: fallback-secret-name
+  namespace: projectcontour
+envoy-client-certificate:
+  name: envoy-client-cert-secret-name
+  namespace: projectcontour
+leaderelection:
+  configmap-name: leader-elect
+  configmap-namespace: projectcontour
+enableExternalNameService: false
+accesslog-format: envoy
+accesslog-format-string: "...\n"
+json-fields:
+  - <Fields Omitted)
+default-http-versions:
+  - "HTTP/2"
+  - "HTTP/1.1"
+timeouts:
+  request-timeout: infinity
+  connection-idle-timeout: 60s
+  stream-idle-timeout: 5m
+  max-connection-duration: infinity
+  delayed-close-timeout: 1s
+  connection-shutdown-grace-period: 5s
+cluster:
+  dns-lookup-family: auto
+network:
+  num-trusted-hops: 0
+  admin-port: 9001
+rateLimitService:
+  extensionService: projectcontour/ratelimit
+  domain: contour
+  failOpen: false
+  enableXRateLimitHeaders: false
+policy:
+  request-headers:
+    set:
+  response-headers:
+    set:
+```
+
+The contents of this file has grown over time and some fields need to be better categorized. 
+The new ContourConfiguration CRD will move the logging options into a `logging` struct. 
+All other fields will stay the same where they currently exist in the configmap.
+
+### New fields:
+
+Some fields will be moved around and others added to complete the configuration:
```

Comment:
Would it be appropriate to move even more `contour serve` command line flags to config CRD?

There could be `spec.metrics` category with `spec.metrics.contour` and `spec.metrics.envoy`. I was also wishing to add certificates for metrics #3707, which would add server certificate, private key and CA certificate for both contour and envoy. Current command line flags

```
--http-address=<ipaddr>    Address the metrics HTTP endpoint will bind to.
--http-port=<port>         Port the metrics HTTP endpoint will bind to.
--stats-address=<ipaddr>   Envoy /stats interface address.
--stats-port=<port>        Envoy /stats interface port.
```

Though maybe same could be asked for some other parameters as well such as envoy HTTP/HTTPS access log file paths, proxy protocol...

Reply:
I think these are fine to move in, we will need to keep the command line args around for a while and do a proper deprecation pass on them though.

In general, I think that we should keep command line flags to a minimum, and to things that are difficult to change once the pod has started (I'm thinking the Contour log file destination argument that was recently proposed is a good example here).

Reply:
I need to double check again, but I think I've gotten all the `contour serve` flags moved into the config file.